### DISCUSSION TITLE
Using dhall to modify container image definition

### DISCUSSION THREAD

**Discussion by JayDoubleu** [podenv/podenv#71] [General]
Hi, Thanks for your help in https://github.com/podenv/podenv/issues/64
I'm new to dhall so still trying to get my head around it.

Would you be able to advise on the recommended way of extending existing image without the need for extra argument ? 

Currently I'm getting around it by specifying empty `RUN echo` as shown below:
```dhall
let Podenv = env:PODENV

let Hub = Podenv.Hub

let container =
      let extra =
            ''
            RUN echo
            ''

      in  \(packages : List Text) ->
            Podenv.Container (Hub.fedora.useImage "latest" extra packages)

in  Podenv.Application::{
    , description = Some "Gnome tweaks"
    , runtime =
        container
          [ "gnome-tweaks"]
    , command = [ "gnome-tweaks" ]
    , capabilities = Podenv.Capabilities::{
      , dbus = True
      , dri = True
      , pipewire = True
      , wayland = True
      }
    }
```

But when I remove the extra bit it throws:
```
  |
6 |       in  \(packages : List Text) ->
  |         ^
expecting expression or whitespace
```

Also it would be great to understand how to extend it multiple times AKA:
- run some commands via `RUN command`
- Grab latest image and install additional packages
- run some commands via `RUN command` again.

**Comment by TristanCacqueray** [comment]
To fix your immediate issue I suggest you use `runtime = Hub.fedora.latest.use ["gnome-tweaks"]` in the Application record. If you have the `dhall` command installed, you can see the result by running in bash:

```ShellSession
env PODENV="$(podenv --dhall-env)" dhall <<< '(env:PODENV).Hub.fedora.latest.use ["gnome-tweaks"]'
```

... which should display:

```dhall
< Container :
    { containerfile : Text
    , image_home : Optional Text
    , image_name : Optional Text
    , image_update : Optional Text
    , image_volumes : List Text
    }
| Image : Text
| Nix : Text
>.Container
  { containerfile =
      ''
      FROM registry.fedoraproject.org/fedora:latest
      ARG USER_UID
      RUN useradd -u ''${USER_UID} -d /home/fedora -m fedora && mkdir -p /home/fedora/.cache /home/fedora/.config /run/user/''${USER_UID} /run/user/0 && chown -R ''${USER_UID}:''${USER_UID} /home/fedora /run/user/

      RUN dnf update -y
      RUN dnf install -y gnome-tweaks
      ''
  , image_home = Some "/home/fedora"
  , image_name = None Text
  , image_update = Some "dnf update -y"
  , image_volumes = [ "cache-dnf-latest:/var/cache/dnf" ]
  }
```

Now to understand how to tweak the runtime, here are some notes about the dhall language:

## let bindings

The error you get is because `in ...` is not a valid expression, it can only be used after `let` definitions.
In dhall (and a few other languages like OCaml or Haskell), you can give a name to an expression using let bindings. This is useful to split a big expression in smaller pieces. The syntax is: `let name = expr in final-expression`, for example:

```dhall
let greeting = "Hello"

let name = "alice"

in greeting ++ " " ++ name
```

The whole `let .. in ...` syntax also forms an expression, so you can nest them, for example:

```dhall
let gretting =
  let name = "alice"
  in "Hello " ++ name
let mark = "!"
in greeting ++ mark
```

... which is equivalent to `let greeting = "Hello alice" let mark = "!" in greeting ++ mark` which is `"Hello alice!"`

## Imports

Expression can be imported using url, environment or files. When using podenv, the `env:PODENV` is set the hub package. You can actually see the value by running:

```ShellSession
$ podenv --dhall-env
https://raw.githubusercontent.com/podenv/hub/41c0711660bb9baa169f761a48bc04d23a4f484b/package.dhall sha256:<semantic-hash>
```

It is quite convenient to be able to follow how the imports are resolved, so let's track where `Hub.fedora` is comming from:

So `env:PODENV` is `https://github.com/podenv/hub/blob/main/package.dhall`
There you can see that this is a record with an attribute `Hub` pointing at `https://github.com/podenv/hub/blob/main/Applications/package.dhall`
Which is another record with a `fedora` attribute. Thus when you write `(env:PODENV).Hub.fedora` you are actually accessing `https://github.com/podenv/hub/blob/main/Applications/fedora.dhall`

## Functions

`fedora.latest.use` is defined here https://github.com/podenv/hub/blob/6dab4d345e5efd663b30a653857e5dbc49513aa5/Applications/fedora.dhall#L39
and it is using the `mkUse` function defined at the top of the same file.

`mkUse` is a function that expects three arguments, and `fedora.latest.use` provides the first two (version set to "latest", and pre-tasks set to "").
The result is a function that expect the last argument, that is a list of packages.
This works thanks to something called [currying](https://en.wikipedia.org/wiki/Currying), or function pre-application. Here is a simpler example to see what is going on:

```dhall
-- | add is a function that takes a natural `x` and returns a function that takes another natural `y`
--   that will return `x + y`
let add = \(x : Natural) -> \(y : Natural) -> x + y

-- | add2 provides the first argument of add, resulting in a function that expects a natural `y`, and
--   add2 will returns `2 + y`
let add2 = add 2

in add2 40
``` 

... which is equivalent to `42`.

So when you write `fedora.latest.use ["gnome-tweaks"]` this is equivalent to `mkUse "latest" "" ["gnome-tweaks"]` which is equivalent to `Podenv.Container (fedora.image "latest" "" ["gnome-tweaks"])`

And here is the actual definition of `fedora.image`: https://github.com/podenv/hub/blob/6dab4d345e5efd663b30a653857e5dbc49513aa5/Builders/fedora.dhall


As you can see in that last link, the image definition is rather simple, it just let you run extra steps before doing the package install, which happens to be enough for the current hub usage.

I am not sure what you mean by `extend it multiple times`, but that would need to happen in that `Builders/fedora.dhall` file. Also note that you can use a completely different function to create the `runtime` attribute of your application

**Comment by TristanCacqueray** [comment]
Another solution for your example is to use the `useSimple` function, which works well when the package name is the command line name. Thus your example can be rewritten as:

```dhall
(env:PODENV).Hub.fedora.useSimple "gnome-tweaks" "Gnome tweaks"
with capabilities = {
      , dbus = True
      , dri = True
      , pipewire = True
      , wayland = True
      }
```

Here is a bit more information about dhall record:

## record with attribute = value

It is often necessary to override an attribute of a record, for example to set the capabilities of an existing application.
In the previous example, `useSimple` returns a complete Application (`use` returns just the runtime). Here is a more simple example:

```dhall
let person = { name = "alice", friend = "bob" }

in person with friend = "eve"
```

... is equivalent to `{ name = "alice", friend = "eve" }`

## record1 `//` record2

Another option is to use `//` which merge two records with the left one taking the priority. Thus the simple example can be rewritten as:

```dhall
let person = { name = "alice", friend = "bob" }

in person // { friend = "eve" }
```

**Reply by JayDoubleu** [reply]
Wow, thanks for a very comprehensive breakdown of this, it helps a lot !

What I mean by extending it is that I would like to run some commands before and after packages are installed:
AKA:

```dhall
let container =
      let extra_before =
            ''
            RUN command 1
            ''
      let extra_after =
            ''
            RUN command 2
            ''
      in  \(packages : List Text) ->
            Podenv.Container (Hub.fedora.useImage "latest" extra_before packages extra_after)

```
This should result in `command 1` running first, then it should install extra packages after which should run `command 2`

**Comment by TristanCacqueray** [comment]
You are welcome, I hope this is useful :-)

Ok I understand what you want. So `useImage` produces a record defined as: https://github.com/podenv/hub/blob/main/schemas/ContainerBuild.dhall.

Thus we can define a function to modify that record. For example here is a function that takes an extra_after Text command, an image ContainerBuild record, and it returns a new ContainerBuild:

```dhall
\(extra_after : Text) ->
\(image : (env:Podenv).ContainerBuild.Type) ->
  image with containerfile = image.containerfile ++ extra_after ++ "\n"
``` 

Which you can use like this:

```dhall
let Podenv = env:PODENV

let add_final_task =
      \(extra_after : Text) ->
      \(image : Podenv.ContainerBuild.Type) ->
        image with containerfile = image.containerfile ++ extra_after ++ "\n"

let container =
      \(packages : List Text) ->
        let extra_before =
              ''
              RUN command 1
              ''

        let extra_after =
              ''
              RUN command 2
              ''

        let image = Podenv.Hub.fedora.useImage "latest" extra_before packages

        in  Podenv.Container (add_final_task extra_after image)

in  Podenv.Application::{
    , description = Some "Gnome tweaks"
    , runtime = container [ "gnome-tweaks" ]
    , command = [ "gnome-tweaks" ]
    , capabilities = Podenv.Capabilities::{
      , dbus = True
      , dri = True
      , pipewire = True
      , wayland = True
      }
    }
```

And if you put that expression in a file named `tweaks.dhall`, running `podenv --config ./tweaks.dhall --show` display:

```
# Containerfile localhost/
FROM registry.fedoraproject.org/fedora:latest
ARG USER_UID
RUN useradd -u ${USER_UID} -d /home/fedora -m fedora && mkdir -p /home/fedora/.cache /home/fedora/.config /run/user/${USER_UID} /run/user/0 && chown -R ${USER_UID}:${USER_UID} /home/fedora /run/user/
RUN command 1

RUN dnf update -y
RUN dnf install -y gnome-tweaks
RUN command 2
```

**Comment by TristanCacqueray** [comment]
For completeness, note that you can also provide your own ContainerBuild record without using the one provided by the fedora `useImage` function. Here is another definition of the `tweaks.dhall` shown previously:

```dhall
let Podenv = env:PODENV

let Text/concatSep =
      https://prelude.dhall-lang.org/v17.0.0/Text/concatSep sha256:e4401d69918c61b92a4c0288f7d60a6560ca99726138ed8ebc58dca2cd205e58

let container =
      \(packages : List Text) ->
        let extra_before =
              ''
              RUN command 1
              ''

        let extra_after =
              ''
              RUN command 2
              ''

        let image =
              Podenv.ContainerBuild::{
              , containerfile =
                  ''
                  FROM registry.fedoraproject.org/fedora:latest
                  ARG USER_UID
                  RUN useradd -u ''${USER_UID} -d /home/fedora -m fedora && mkdir -p /home/fedora/.cache /home/fedora/.config /run/user/''${USER_UID} /run/user/0 && chown -R ''${USER_UID}:''${USER_UID} /home/fedora /run/user/
                  ${extra_before}
                  RUN dnf update -y
                  RUN dnf install -y ${Text/concatSep " " packages}
                  ${extra_after}
                  ''
              , image_home = Some "/home/fedora"
              , image_update = Some "dnf update -y"
              , image_volumes = [ "cache-dnf-latest:/var/cache/dnf" ]
              }

        in  Podenv.Container image

in  Podenv.Application::{
    , description = Some "Gnome tweaks"
    , runtime = container [ "gnome-tweaks" ]
    , command = [ "gnome-tweaks" ]
    , capabilities = Podenv.Capabilities::{
      , dbus = True
      , dri = True
      , pipewire = True
      , wayland = True
      }
    }
``` 

As a matter of fact, you can use dhall semantic hash to verify the two versions are identicals, for example running:

```ShellSession
env PODENV="$(podenv --dhall-env)" dhall hash --file ./tweaks.dhall
```

... shall display `sha256:98077fa78cc08cbb2c07739d04c7cbc22849f78360bcc59ad2409cadd0b2d7a9` for both versions.
